plugins/common/opcua/input/input_client.go

```diff
@@ -0,0 +1,394 @@
+package input
+
+import (
+	"context"
+	"fmt"
+	"github.com/gopcua/opcua/ua"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal/choice"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/common/opcua"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type TimestampSource string
+
+const (
+	TimestampSourceServer   TimestampSource = "server"
+	TimestampSourceSource   TimestampSource = "source"
+	TimestampSourceTelegraf TimestampSource = "gather"
+)
+
+// NodeSettings describes how to map from a OPC UA node to a Metric
+type NodeSettings struct {
+	FieldName      string            `toml:"name"`
+	Namespace      string            `toml:"namespace"`
+	IdentifierType string            `toml:"identifier_type"`
+	Identifier     string            `toml:"identifier"`
+	DataType       string            `toml:"data_type" deprecated:"1.26.0;not used"`
+	Description    string            `toml:"description" deprecated:"1.26.0;not used"`
+	TagsSlice      [][]string        `toml:"tags" deprecated:"1.26.0;use default_tags"`
```

Comment:
Looks better, only they are not used since at least telegraf 1.17 and the next feature release will be 1.25 mid december.
```suggestion
	DataType       string            `toml:"data_type" deprecated:"1.17.0;not used"`
	Description    string            `toml:"description" deprecated:"1.17.0;not used"`
	TagsSlice      [][]string        `toml:"tags" deprecated:"1.25.0;use default_tags"`
```

Reply:
All looks good, don't forget to update sample.conf and readme documentation..

Reply:
I'm trying to update the config reader test with the following input:
```toml
[[inputs.opcua]]
name = "localhost"
endpoint = "opc.tcp://localhost:4840"
connect_timeout = "10s"
request_timeout = "5s"
security_policy = "auto"
security_mode = "auto"
certificate = "/etc/telegraf/cert.pem"
private_key = "/etc/telegraf/key.pem"
auth_method = "Anonymous"
username = ""
password = ""
nodes = [
  {name="name",  namespace="1", identifier_type="s", identifier="one"},
  {name="name2", namespace="2", identifier_type="s", identifier="two"},
]

[[inputs.opcua.group]]
name = "foo"
namespace = "3"
identifier_type = "i"
tags = [["tag1", "val1"], ["tag2", "val2"]]
nodes = [{name="name3", identifier="3000", tags=[["tag3", "val3"]]}]

[[inputs.opcua.group]]
name = "bar"
namespace = "0"
identifier_type = "i"
tags = [["tag1", "val1"], ["tag2", "val2"]]
nodes = [{name="name4", identifier="4000", default_tags={ tag1 = "override" } }]

[inputs.opcua.workarounds]
additional_valid_status_codes = ["0xC0"]

[inputs.opcua.request_workarounds]
use_unregistered_reads = true
```

However, I get the following error 
`error parsing opcua, line 31: (input.NodeSettings.DefaultTags) cannot unmarshal TOML array table into map[string]string (need slice)`

I am not entirely sure why it is not working. It is valid TOML, but the unmarshaling seems to fail. This might be the original reason for using array slices, but I'd rather solve the root problem.

If I put in the following configuration, it works, but it is very ugly and unwieldy:

```toml
[[inputs.opcua]]
name = "localhost"
endpoint = "opc.tcp://localhost:4840"
connect_timeout = "10s"
request_timeout = "5s"
security_policy = "auto"
security_mode = "auto"
certificate = "/etc/telegraf/cert.pem"
private_key = "/etc/telegraf/key.pem"
auth_method = "Anonymous"
username = ""
password = ""
nodes = [
  {name="name",  namespace="1", identifier_type="s", identifier="one"},
  {name="name2", namespace="2", identifier_type="s", identifier="two"},
]

[[inputs.opcua.group]]
name = "foo"
namespace = "3"
identifier_type = "i"
tags = [["tag1", "val1"], ["tag2", "val2"]]
nodes = [{name="name3", identifier="3000", tags=[["tag3", "val3"]]}]

[[inputs.opcua.group]]
name = "bar"
namespace = "0"
identifier_type = "i"
tags = [["tag1", "val1"], ["tag2", "val2"]]
nodes = [{name="name4", identifier="4000"}]
[inputs.opcua.group.nodes.default_tags]
tag1 = "override"

[inputs.opcua.workarounds]
additional_valid_status_codes = ["0xC0"]

[inputs.opcua.request_workarounds]
use_unregistered_reads = true
```

Reply:
Hmm, yeah  that is indeed how tags in telegraf config are usually defined.. It might indeed be that there is a limitation/bug in the used TOML parser..

Reply:
It looks like this is the toml parser that is used:
`github.com/influxdata/toml`

How do you want to proceed?

Reply:
Discussed this with the team, as this is a highly wanted feature we would go with the ugly way first to get this merged.

Do you think you would be able to work on fixing the parser?

Reply:
Ah nice that you guys discussed it. 
What do you mean exactly with the ugly way?
Keep using the string slices or use the verbose syntax for the dictionary?

Unfortunately, I won't have time to look into the parser.

Reply:
I mean to use this way:
```toml
[inputs.opcua.group.nodes.default_tags]
tag1 = "override"
```
So that is how it is right now I guess?

Reply:
Unfortunately, the syntax does not actually work. The validator says it's valid, but that's just the syntax, not the semantics. This means every node must be specified verbosely, instead of inline if there are `default_tags`. Unless, users don't want to use custom tags per node.

The syntax will be like this:
```toml
[[inputs.opcua]]
name = "localhost"
endpoint = "opc.tcp://localhost:4840"
connect_timeout = "10s"
request_timeout = "5s"
security_policy = "auto"
security_mode = "auto"
certificate = "/etc/telegraf/cert.pem"
private_key = "/etc/telegraf/key.pem"
auth_method = "Anonymous"
username = ""
password = ""

[[inputs.opcua.nodes]]
  name = "name"
  namespace = "1"
  identifier_type = "s"
  identifier="one"
  tags=[["tag0", "val0"]]

[[inputs.opcua.nodes]]
  name="name2"
  namespace="2"
  identifier_type="s"
  identifier="two" 
  tags=[["tag0", "val0"], ["tag00", "val00"]]
  default_tags = {tag6 = "val6"}

[[inputs.opcua.group]]
name = "foo"
namespace = "3"
identifier_type = "i"
tags = [["tag1", "val1"], ["tag2", "val2"]]
nodes = [{name="name3", identifier="3000", tags=[["tag3", "val3"]]}]

[[inputs.opcua.group]]
name = "bar"
namespace = "0"
identifier_type = "i"
tags = [["tag1", "val1"], ["tag2", "val2"]]
[[inputs.opcua.group.nodes]]
  name = "name4"
  identifier = "4000"
  tags=[["tag4", "val4"]]
  default_tags = { tag1 = "override" }

[inputs.opcua.workarounds]
additional_valid_status_codes = ["0xC0"]

[inputs.opcua.request_workarounds]
use_unregistered_reads = true
```

The following series keys will be output:
* `localhost,tag0=val0 name`
* `localhost,tag6=val6 name2`
* `foo,tag1=val1,tag2=val2,tag3=val3 name3`
* `bar,tag1=override,tag2=val2 name4`

Reply:
Yes correct, your previous example wasn't also valid TOML according to linters. If you want further discussion on this, we could do this in  the main thread instead of here (here I'm  discussing a change to the deprecation tags)

Reply:
It's clear now, I will adapt the test cases.